### DISCUSSION TITLE
New track changes gutter band

### DISCUSSION THREAD

**Discussion by pyscripter** [TurboPack/SynEdit#122] [Announcements]
The repository now includes the  implementation of a track changes  gutter band similar to  the on found in Visual Studio (see [here](https://dailydotnettips.com/color-indicator-for-code-changes-track-changes-in-visual-studio/)).

The [main SynEdit project](https://github.com/SynEdit/SynEdit) does have an implementation, but line status is not updated during undo/redo.  

### How to use it

Gutter contains a new published property TrackChanges which controls the visibility and appearance of the Track Changes getter band.  This property type is TSynTrackChanges defined as follows;

 ```pascal
TSynTrackChanges = class(TPersistent)
  ...
  published
    property Width: Integer read FWidth write FWidth default 4;
    property Visible: Boolean read FVisible write FVisible default False;
    property SavedColor: TColor read FSavedColor write FSavedColor
      default clGreen;
    property ModifiedColor: TColor read FModifiedColor write FModifiedColor
      default clYellow;
    property SavedModifiedColor: TColor read FSavedModifiedColor
      write FSavedModifiedColor default clWebOrange;
    property OriginalColor: TColor read FOriginalColor write FOriginalColor
      default clBlue;
  end;
```
The constructor of TSynGutter creates a band of gbkTrackChanges kind, which by default is not visible.

### Backward compatibility note
You need to manually add this band in existing forms that contain Synedit .

### What it looks like

a) Displayed with the editor background

![image](https://user-images.githubusercontent.com/1311616/145921343-38bb3e26-4c32-4d3c-a405-e68c2b5e81c8.png)

b) Displayed with the gutter background

![image](https://user-images.githubusercontent.com/1311616/145921789-fd5d9167-ca85-4d2b-b740-f70a59724e2b.png)

### Meaning of the colors (customizable)
1. Yellow: modified line
2. Green: line modified and saved.
3. Orange: line modified, saved and then modified again
4. Blue: line modified, saved but then the changes were undone and it is now as in the original version.

###  Related SynEdit methods
 ```pascal
    //  Clears track change info
    procedure ClearTrackChanges;
    // marks the text buffer as saved.  Call when you save the text buffer
    procedure MarkSaved;
```
Note that ClearUndo calls ClearTrackChanges

@JaFi-cz @MShark67 Please test and provide feedback.

**Comment by JaFi-cz** [comment]
Great. Thanks

**Comment by JaFi-cz** [comment]
Just testing and can't get any other color than yellow.
When I change any content, yellow color appear. OK.
Undo resets color back to none.
When I save file, it stays yellow. But I don't use SynEdit.SaveToFile() method, cause I process file with plugin, change code page e.t.c. I use Synedit.Lines.SaveToStream() and change Modified to False.
I need to investigate what is necessary to make the band colorfull  ;-)

**Reply by MShark67** [reply]
Note that you do need to call MarkSaved; after you do the save.  I do get all the colors here.

**Comment by JaFi-cz** [comment]
Thank you, this was the point. Now it works all ok.

**Comment by MShark67** [comment]
Looks very good!  I think that the SynEdit.Gutter.TrackChanges.Visible property needs a setter that calls OnChanged.  Right now I need to toggle Gutter.Visible in order to see the change.

-Mark

**Comment by pyscripter** [comment]
@MShark67
Created issue #123 and fixed.   Please check.

**Comment by pyscripter** [comment]
Also changed default colors. What do you think?

**Reply by MShark67** [reply]
New colors are good, especially the "yellow".  It's a bit easier on the eyes for me.

**Reply by JaFi-cz** [reply]
I agree, new colors are better

**Comment by sllimv1973** [comment]
Thanks for this, I've been carrying those modifications forward for years.  It's about time they made the mainline code.  Cheers.
